### ISSUE TITLE
inference single image

### ISSUE THREAD

**Issue by SatMa34** [open-mmlab/mmpose#582]
Hi, do you have the api that i can inference one single image? the input is one single image, and the output is the keypoints. I notice you have demo/inference_bottom_up_pose_model.py, but there is a arg named json_file, what dose this arg mean? In inference mode, the information all we have is the image itself, but in the demo/inference_bottom_up_pose_model.py, it uses the gt information, it's not a stardard inference process.

**Comment by innerlee** [comment]
Try demos with mmdet

**Comment by jin-s13** [comment]
Only `file_name` in the `json_file` is used. We do not use any gt information.

**Comment by jin-s13** [comment]
To inference a single image, you need to prepare a json file like this:

```
{
    "images": [
        {
            "width": 1239,
            "height": 731,
            "file_name": "003464.jpg",
            "id": 3464
        }
    ]
}
```
Note that we only need `images`.

**Comment by MIMOS-Chien** [comment]
Hi, I am also trying to run  "demo/top_down_img_demo_with_mmdet.py". But failed to run this program. It requires mmdet installation. I have installed mmpose. Should I install mmdet in the same env or create a new env out of the mmpose env? After installation how to run that? TQ

**Comment by jin-s13** [comment]
You should install mmdet in the same environment.

**Comment by MIMOS-Chien** [comment]
It seems that it doesn't work for me even I have installed mmdet.


File "C:/mmlab/mmpose-master/demo/top_down_img_demo_with_mmdet.py", line 136, in <module>
    main()
  File "C:/mmlab/mmpose-master/demo/top_down_img_demo_with_mmdet.py", line 75, in main
    assert has_mmdet, 'Please install mmdet to run the demo.'
AssertionError: Please install mmdet to run the demo.


(c:\mmlab\open-mmlab) C:\mmlab\mmpose-master\mmdetection>conda list
# packages in environment at c:\mmlab\open-mmlab:
#
# Name                    Version                   Build  Channel
addict                    2.4.0                    pypi_0    pypi
:
:
cudatoolkit               10.1.243             h74a9793_0
cycler                    0.10.0                   pypi_0    pypi
:
:
mmcv                      1.3.1                     dev_0    <develop>
mmdet                     2.10.0                    dev_0    <develop>



(c:\mmlab\open-mmlab) C:\mmlab\mmpose-master\mmdetection>pip list
Package                       Version             Location
----------------------------- ------------------- -----------------------------------------------------------
addict                        2.4.0
:
:
coverage                      5.5
:
:
mmcv                          1.3.0               c:\users\user\appdata\roaming\python\python38\site-packages
mmdet                         2.10.0              c:\mmlab\mmpose-master\mmdetection

**Comment by ly015** [comment]
@MIMOS-Chien Could you please try importing mmdet to see if there is any error information? If it's something related to mmcv, maybe installing mmcv-full instead of mmcv can solve this problem.

**Comment by MIMOS-Chien** [comment]
I didn't manage to install via method  {pip install mmcv-full -f https://download.openmmlab.com/mmcv/dist/cu110/torch1.7.1/index.html} even I have tried many times. I only managed to install mmcv by downloading the module and run "python setup.py develop". It installs version mmcv==1.3.1 for me. 

But now, after few attempts of installations/uninstallations, suddenlyy this trouble comes out "No CUDA runtime is found, using CUDA_HOME='C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2'
OMP: Error #15: Initializing libiomp5md.dll, but found libiomp5md.dll already initialized." Any suggestions? TQ

**Comment by MIMOS-Chien** [comment]
{git clone https://github.com/open-mmlab/mmcv.git
cd mmcv
MMCV_WITH_OPS=1 pip install -e .  # package mmcv-full will be installed after this step
cd ..}

Can I go to the file change  {name='mmcv' if os.getenv('MMCV_WITH_OPS', '0') == '0' else 'mmcv-full'}
     to {name='mmcv-full' if os.getenv('MMCV_WITH_OPS', '1') == '1' else 'mmcv'}. 

Anyhow, the biggest problem now has shifted to "No CUDA runtime is found, using CUDA_HOME=....". Even troblesome

**Comment by MIMOS-Chien** [comment]
cls

After reinstall "pip install torch==1.8.1+cu102 torchvision==0.9.1+cu102 torchaudio===0.8.1 -f https://download.pytorch.org/whl/torch_stable.html",  "No CUDA runtime is found, using CUDA_HOME=...." no longer exist. BUT, " OMP: Error #15: Initializing libiomp5md.dll, but found libiomp5md.dll already initialized" still exist.

**Comment by MIMOS-Chien** [comment]
Hi, I am still stuck at "assert has_mmdet, 'Please install mmdet to run the demo.'
AssertionError: Please install mmdet to run the demo"

According to the installation guide, we create a new environment "open-mmlab". Then install other necessary pakages. When reaching the stage for installing MMDetection, we make a clone of the module and go into the folder, install it later. I managed to install it. (mmdet                     2.10.0                   pypi_0    pypi). What is the reason that it doesn't recognize mmdet even it has already installed? Kindly advise. TQ

**Comment by ly015** [comment]
Could you please add a line "raise" after [L11](https://github.com/open-mmlab/mmpose/blob/master/demo/top_down_img_demo_with_mmdet.py#L11) in demo/top_down_img_demo_with_mmdet.py to expose the error information which may help us locate the problem?

**Comment by MIMOS-Chien** [comment]
print("Before mmdet verification")
try:
    from mmdet.apis import inference_detector, init_detector
    has_mmdet = True
except (ImportError, ModuleNotFoundError):
    has_mmdet = False

print("after mmdet verification") 

OUTPUT:
Before mmdet verification
after mmdet verification
Traceback (most recent call last):
  File "C:/mmlab/mmpose-master/demo/top_down_img_demo_with_mmdet.py", line 138, in <module>
    main()
  File "C:/mmlab/mmpose-master/demo/top_down_img_demo_with_mmdet.py", line 77, in main
    assert has_mmdet, 'Please install mmdet to run the demo.'
AssertionError: Please install mmdet to run the demo.

**Comment by ly015** [comment]
> print("Before mmdet verification")
> try:
> from mmdet.apis import inference_detector, init_detector
> has_mmdet = True
> except (ImportError, ModuleNotFoundError):
> has_mmdet = False
> 
> print("after mmdet verification")
> 
> OUTPUT:
> Before mmdet verification
> after mmdet verification
> Traceback (most recent call last):
> File "C:/mmlab/mmpose-master/demo/top_down_img_demo_with_mmdet.py", line 138, in
> main()
> File "C:/mmlab/mmpose-master/demo/top_down_img_demo_with_mmdet.py", line 77, in main
> assert has_mmdet, 'Please install mmdet to run the demo.'
> AssertionError: Please install mmdet to run the demo.

The error information of importing mmdet is suppressed by try-except statement. Please add a `raise` line after L11, or replace L7-L11 with a single line `from mmdet.apis import inference_detector, init_detector`, to expose the exception thrown from mmdet importing.

**Comment by MIMOS-Chien** [comment]
Dear all, I think I  almost spent about 7 days for installing mmdect + mmcv-full. mmcv-full is a major problem.  I follow the link below:

https://www.zywvvd.com/2021/04/20/deep_learning/windows-mmcv-1-2-7-install/windows-mmcv-1-2-7-install/

Thank you for your kind assistance. 
One of the important obstacles is the cl.exe setting in the "PATH".

By running the op_down_img_demo_with_mmdet.py again as follows 

(c:\mmlab\open-mmlab) C:\mmlab\mmdetection>python C:/mmlab/mmpose-master/demo/top_down_img_demo_with_mmdet.py C:/mmlab/mmpose-master/demo/mmdetection_cfg/cascade_rcnn_x101_64x4d_fpn_1class.py C:/mmlab/mmpose-master/path/cascade_rcnn_x101_64x4d_fpn_20e_onehand10k-dac19597_20201030.pth C:/mmlab/mmpose-master/configs/hand/resnet/onehand10k/res50_onehand10k_256x256.py C:/mmlab/mmpose-master/path/res50_onehand10k_256x256-e67998f6_20200813.pth --img-root C:/mmlab/mmpose-master/tests/data/onehand10k/ --img 9.jpg --out-img-root A10



c:\mmlab\mmdetection\mmdet\models\builder.py:72: UserWarning: train_cfg and test_cfg is deprecated, please specify them in model
  'please specify them in model', UserWarning)
Use load_from_local loader
Use load_from_local loader
c:\mmlab\mmdetection\mmdet\datasets\utils.py:68: UserWarning: "ImageToTensor" pipeline is replaced by "DefaultFormatBundle" for batch inference. It is recommended to manually replace it in the test data pipeline in your config file.
  'data pipeline in your config file.', UserWarning)
c:\mmlab\mmdetection\mmdet\models\dense_heads\rpn_head.py:192: UserWarning: In rpn_proposal or test_cfg, nms_thr has been moved to a dict named nms as iou_threshold, max_num has been renamed as max_per_img, name of original arguments and the way to specify iou_threshold of NMS will be deprecated.
  'In rpn_proposal or test_cfg, '

What should I do for these since there are not many useful information from the web? By right, I should got output in the folder A10, right? It is empty without any error messages but with warning described above. TQ

**Comment by ly015** [comment]
These warnings are caused by out-of-date mmdetection configs used in the demo (fixed in #592). But these configs should still be able to work, so it probably has nothing to do with this issue. Could you please check that the image path is correct and the image has been successfully loaded?
